pkg/listwatch/listwatch.go

```diff
@@ -0,0 +1,214 @@
+		// Since the client does not have Watch privileges, do not
+		// actually watch anything. Use a watch.FakeWatcher here to
+		// implement watch.Interface but not send any events.
+		return watch.NewFake(), nil
+	}
+	return &cache.ListWatch{ListFunc: listFunc, WatchFunc: watchFunc}
+}
+
+// MultiNamespaceListerWatcher takes a list of namespaces and a
+// cache.ListerWatcher generator func and returns a single cache.ListerWatcher
+// capable of operating on multiple namespaces.
+func MultiNamespaceListerWatcher(namespaces []string, f func(string) cache.ListerWatcher) cache.ListerWatcher {
+	// If there is only one namespace then there is no need to create a
+	// proxy.
+	if len(namespaces) == 1 {
+		return f(namespaces[0])
+	}
+	var lws []cache.ListerWatcher
+	for _, n := range namespaces {
+		lws = append(lws, f(n))
+	}
+	return multiListerWatcher(lws)
+}
+
+// multiListerWatcher abstracts several cache.ListerWatchers, allowing them
+// to be treated as a single cache.ListerWatcher.
+type multiListerWatcher []cache.ListerWatcher
+
+// List implements the ListerWatcher interface.
+// It combines the output of the List method of every ListerWatcher into
+// a single result.
+func (mlw multiListerWatcher) List(options metav1.ListOptions) (runtime.Object, error) {
+	l := metav1.List{}
+	var resourceVersions []string
+	for _, lw := range mlw {
+		list, err := lw.List(options)
+		if err != nil {
+			return nil, err
+		}
+		items, err := meta.ExtractList(list)
+		if err != nil {
+			return nil, err
+		}
+		metaObj, err := meta.ListAccessor(list)
+		if err != nil {
+			return nil, err
+		}
+		for _, item := range items {
+			l.Items = append(l.Items, runtime.RawExtension{Object: item.DeepCopyObject()})
+		}
+		resourceVersions = append(resourceVersions, metaObj.GetResourceVersion())
+	}
+	// Combine the resource versions so that the composite Watch method can
+	// distribute appropriate versions to each underlying Watch func.
+	l.ListMeta.ResourceVersion = strings.Join(resourceVersions, "/")
+	return &l, nil
+}
+
+// Watch implements the ListerWatcher interface.
+// It returns a watch.Interface that combines the output from the
+// watch.Interface of every cache.ListerWatcher into a single result chan.
+func (mlw multiListerWatcher) Watch(options metav1.ListOptions) (watch.Interface, error) {
+	resourceVersions := make([]string, len(mlw))
+	// Allow resource versions to be "".
+	if options.ResourceVersion != "" {
+		rvs := strings.Split(options.ResourceVersion, "/")
+		if len(rvs) != len(mlw) {
+			return nil, fmt.Errorf("expected resource version to have %d parts to match the number of ListerWatchers", len(mlw))
+		}
+		resourceVersions = rvs
+	}
+	return newMultiWatch(mlw, resourceVersions, options)
+}
+
+// multiWatch abstracts multiple watch.Interface's, allowing them
+// to be treated as a single watch.Interface.
+type multiWatch struct {
+	result   chan watch.Event
+	stopped  bool
+	stoppers []func()
+	sync.Mutex
```

Comment:
this embeds `sync.Mutex` into `multiWatch`, effectively exporting `Lock()` und `Unlock()` to the outside world of but you are using those methods only internally (which is good).

Can we make a private `mu sync.Mutex // protects the stopped field only` so those methods don't escape and consumers of `multiWatch` don't even try to lock/unlock externally?

Reply:
`multiWatch` is an unexported type so only this package can access `Lock` and `Unlock`. Furthermore, anything in this package has access to unexported struct fields, so there is no added security, no?

Embedding the struct this way was mainly for convenience so if you feel strongly, I can make the mutex private.

Reply:
Agreed with @s-urbaniak

Reply:
:+1: